### PR TITLE
Store both the rounded and unrounded node size in Node

### LINKED ISSUE #9874: Text wraps incorrectly for a frame after a change to its font size
## Bevy version

main 038d113

## What you did

```
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .add_systems(Update, update)
        .run();
}

fn setup(
    mut commands: Commands,
) {
    commands.spawn(Camera2dBundle::default());
    commands.spawn(TextBundle::from("Hello, world!"));
}

fn update(
    mut text_query: Query<&mut Text>,
) {
    for mut text in text_query.iter_mut() {
        text.sections[0].style.font_size *= 1.001;
    }
}
```

## What went wrong

As the text increases in size it flickers between being displayed on one line and wrapped across two lines
It should always be on one line.

## Additional Information

Could just be a system ordering issue

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

Text bounds are computed by the layout algorithm using the text's measurefunc so that text will only wrap after it's used the maximum amount of available horizontal space.

When the layout size is returned the layout coordinates are rounded and this sometimes results in the final size of the Node not matching the size computed with the measurefunc. This means that the text may no longer fit the horizontal available space and instead wrap onto a new line. However, no glyphs will be generated for this new line because no vertical space for the extra line was allocated.

fixes #9874

## Solution

Store both the rounded and unrounded node sizes in `Node`.

Rounding is used to eliminate pixel-wide gaps between nodes that should be touching edge to edge, but this isn't necessary for text nodes as they don't have solid edges.

## Changelog

* Added the `rounded_size: Vec2` field to `Node`.
* `text_system` uses the unrounded node size when computing a text layout.

### CODE REVIEW

**Review by nicoburns — APPROVED** [review]
This makes sense to me. I think the general principle is:

Rounding (pixel-grid alignment) should never affect layout, and should be for rendering only.

For text, I suspect that might mean passing the unrounded value into text layout, and then clipping to the rounded value when rendering.

**Comment by ickshonpe** [comment]
> This makes sense to me. I think the general principle is:
> 
> Rounding (pixel-grid alignment) should never affect layout, and should be for rendering only.
> 
> For text, I suspect that might mean passing the unrounded value into text layout, and then clipping to the rounded value when rendering.

Yeah it needs a lot of thinking about. I also considered just adding both the rounded and unrounded sizes to Node.  I was feeling before that it would be a bit redundant and a waste of memory, but perhaps it's a better solution.

**Comment by nicoburns** [comment]
In theory, you shouldn't need to perform text layout again once Taffy's layout has completed. But I'm guessing our current text infrastructure doesn't allow us to reuse the result? We ended up having to store both in Taffy, because otherwise we can't cache results from previous layout runs (as they might round differently if their parents lay out differently and they end up in a different absolute position within the viewport).

**Comment by rparrett** [comment]
I tested every example (scale factor 2 here), and this seems to fix text layout problems with quite a few of them and doesn't seem to introduce any new ones.

The problematic ones were
```
display_and_visibility
overflow
text_wrap_debug
overflow_debug
ui
text_debug
```

**Comment by ickshonpe** [comment]
Switched approach to storing both the rounded and unrounded node sizes in `Node`.

**Comment by ickshonpe** [comment]
> I tested every example at scale factor `2.`, and this seems to fix text layout problems with quite a few of them and doesn't seem to introduce any new problems.
> 
> The ones that had differences were
> 
> ```
> display_and_visibility
> overflow
> text_wrap_debug
> overflow_debug
> ui
> text_debug
> ```
> 
> For other scale factors, I only tested the examples listed above.
> 
> At scale factor `4./3.`, this seems to fix some text wrapping issues, but also introduces a 1px gap in `text_wrap_debug`.
> 
> Seemingly the same story at `1.0` -- text layout is better, but 1px gap in `text_wrap_debug`.

I think the gap is gone now with the new approach. Text can use the unrounded bounds for layout calculations but the size of any node containing text will remain rounded so there shouldn't be any gaps in the layout.

File: examples/ui/text_wrap_debug.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@ fn main() {
     App::new()
         .add_plugins(DefaultPlugins)
         .insert_resource(WinitSettings::desktop_app())
+        .insert_resource(UiScale(4. / 3.))
```
Comment (rparrett) [inline examples/ui/text_wrap_debug.rs:11]:
```suggestion
```

**Review by rparrett — APPROVED** [review]
Love how simple this version is.

Did some more testing using the screenshot diffing setup vs the base commit

- All examples at SF 2
- UI examples at SF 1, 4/3, 5/3

Looks good to me. Minor changes in text wrapping behavior that seem more correct, and seemingly no unintended consequences.

**Review by mockersf — APPROVED** [review]

**Pull request merged**